### DISCUSSION TITLE
Copying attributes between spans within a trace

### DISCUSSION THREAD

**Discussion by bmbferreira** [open-telemetry/opentelemetry-collector-contrib#29504] [Q&A]
Hi! I wonder if it would be possible with the opentelemetry collector to copy specific attributes between spans within a single trace. I know we have a processor that allows to group spans by trace but I wonder what should I use to then "propagate" a specific attribute to all the spans within that trace. 

For example I know that some of the spans on the trace will have an attribute named "user.email". I would like to copy that specific attribute to all the spans of the trace. 

Is this possible to do? Thank you.

**Comment by btpka3** [comment]
Using custom SpanProcessor maybe work:

DemoSpanProcessor
```java
package me.test;

import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;
import io.opentelemetry.sdk.trace.data.SpanData;

public class DemoSpanProcessor implements SpanProcessor {
    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        Span parentSpan = Span.fromContext(parentContext);
        if (parentSpan instanceof ReadableSpan) {
            SpanData parentSpanData = ((ReadableSpan) span).toSpanData();
            parentSpanData.getAttributes().forEach((key, value) -> {
                if (span.getAttribute(key) == null) {
                    span.setAttribute((AttributeKey<Object>) key, value);
                }
            });
        }
    }
    @Override
    public boolean isStartRequired() {return true;}
    @Override
    public void onEnd(ReadableSpan span) {}
    @Override
    public boolean isEndRequired() {return false;}
}
```

then use it to init OpenTelemetry :
```java
protected OpenTelemetry newOpenTelemetry() {
    SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
            .addSpanProcessor(new DemoSpanProcessor())
            .build();
    return OpenTelemetrySdk.builder()
            .setTracerProvider(sdkTracerProvider)
            .buildAndRegisterGlobal();
}
```

**Reply by bmbferreira** [reply]
Thanks for sharing. If you're using honeycomb, the original problem might be solved by the relational fields they announced recently: https://www.honeycomb.io/blog/relational-fields-announcement.

With this I don't need to copy attributes to all the spans within a trace because I can just search for the attributes on any span in the same trace.
